File: sdk/python/mitos/pty.py
from __future__ import annotations

import asyncio
import base64
import json
import threading
from typing import Callable, Optional

import websockets  # the async websockets package
from websocket import WebSocketApp  # from the websocket-client package


class PtyHandle:
    """A live interactive pseudo-terminal in a sandbox, mirroring E2B's
    sandbox.pty handle. Output bytes are delivered to on_data on a background
    reader thread; send_input/resize write frames to the guest, kill() force
    closes, and wait() blocks for the exit code.

    The transport is a WebSocket to forkd's GET /v1/pty (subprotocol
    mitos.pty.v1), gated by the per-sandbox bearer token. The token is sent in
    the Authorization header and is never logged.
    """

    def __init__(
        self,
        url: str,
        token: Optional[str],
        on_data: Callable[[bytes], None],
    ):
        self._on_data = on_data
        self._exit_code: Optional[int] = None
        self._done = threading.Event()
        self._open = threading.Event()
        self._lock = threading.Lock()

        header = [f"Authorization: Bearer {token}"] if token else []
        self._ws = WebSocketApp(
            url,
            header=header,
            subprotocols=["mitos.pty.v1"],
            on_open=self._handle_open,
            on_message=self._handle_message,
            on_close=self._handle_close,
            on_error=self._handle_error,
        )
        self._thread = threading.Thread(target=self._ws.run_forever, daemon=True)
        self._thread.start()
        # Block until the socket is open so the first send_input is not dropped.
        self._open.wait(timeout=10)

    def _handle_open(self, ws) -> None:  # noqa: ANN001
        self._open.set()

    def _handle_message(self, ws, message) -> None:  # noqa: ANN001
        frame = json.loads(message)
        kind = frame.get("kind")
        if kind == "output":
            data = frame.get("data")
            decoded = base64.b64decode(data) if data else b""
            self._on_data(decoded)
        elif kind == "exit":
            self._exit_code = int(frame.get("exit_code", 0))
            self._done.set()
            ws.close()

    def _handle_close(self, ws, status_code, msg) -> None:  # noqa: ANN001
        if self._exit_code is None:
            self._exit_code = -1
        self._done.set()

    def _handle_error(self, ws, error) -> None:  # noqa: ANN001
        # Error text never carries the token; record nothing sensitive.
        if self._exit_code is None:
            self._exit_code = -1
        self._done.set()

    def _send(self, frame: dict) -> None:
        with self._lock:
            self._ws.send(json.dumps(frame))

    def send_input(self, data: bytes) -> None:
        """Send raw keystroke bytes to the shell."""
        self._send({"kind": "input", "data": base64.b64encode(data).decode("ascii")})

    def resize(self, cols: int, rows: int) -> None:
        """Resize the terminal window (TIOCSWINSZ in the guest, then SIGWINCH)."""
        self._send({"kind": "resize", "cols": cols, "rows": rows})

    def kill(self) -> None:
        """Force-close the terminal. The guest kills the shell process group
        when the connection drops."""
        try:
            self._ws.close()
        finally:
            self._done.set()

    def wait(self, timeout: Optional[float] = None) -> int:
        """Block until the shell exits and return its exit code (or -1 if the
        connection dropped before a terminal exit frame)."""
        self._done.wait(timeout=timeout)
        return self._exit_code if self._exit_code is not None else -1


class AsyncPtyHandle:
    """Async counterpart to PtyHandle. Output is delivered to on_data from a
    background asyncio task; send_input/resize are coroutines, wait() awaits the
    exit code. Transport is a WebSocket to /v1/pty gated by the bearer token."""

    def __init__(self, ws, on_data: Callable[[bytes], None]):
        self._ws = ws
        self._on_data = on_data
        self._exit_code: Optional[int] = None
        self._done = asyncio.Event()
        self._reader = asyncio.create_task(self._read_loop())

    @classmethod
    async def connect(
        cls,
        url: str,
        token: Optional[str],
        on_data: Callable[[bytes], None],
    ) -> "AsyncPtyHandle":
        headers = [("Authorization", f"Bearer {token}")] if token else []
        ws = await websockets.connect(
            url,
            additional_headers=headers,
            subprotocols=["mitos.pty.v1"],
        )
        return cls(ws, on_data)

    async def _read_loop(self) -> None:
        try:
            async for raw in self._ws:
                frame = json.loads(raw)
                kind = frame.get("kind")
                if kind == "output":
                    data = frame.get("data")
                    self._on_data(base64.b64decode(data) if data else b"")
                elif kind == "exit":
                    self._exit_code = int(frame.get("exit_code", 0))
                    break
        finally:
            if self._exit_code is None:
                self._exit_code = -1
            self._done.set()
            await self._ws.close()

    async def send_input(self, data: bytes) -> None:
        await self._ws.send(
            json.dumps({"kind": "input", "data": base64.b64encode(data).decode("ascii")})
        )

    async def resize(self, cols: int, rows: int) -> None:
        await self._ws.send(json.dumps({"kind": "resize", "cols": cols, "rows": rows}))

    async def kill(self) -> None:
        await self._ws.close()
        self._done.set()

    async def wait(self) -> int:
        await self._done.wait()
        return self._exit_code if self._exit_code is not None else -1

File: sdk/typescript/src/pty.ts
// An interactive pseudo-terminal in a sandbox, mirroring E2B's sandbox.pty.
// Transport is a WebSocket to the sandbox API's /v1/pty (subprotocol
// mitos.pty.v1), gated by the per-sandbox bearer token. The token is sent in
// the Authorization header (the Node `ws` client supports request headers,
// which the platform global WebSocket does not). It is never logged.

import WebSocket from "ws";

/** Frame kinds on the PTY WebSocket. input/resize flow client->guest;
 * output/exit flow guest->client. */
type PtyFrame =
  | { kind: "input"; data: string }
  | { kind: "resize"; cols: number; rows: number }
  | { kind: "output"; data: string }
  | { kind: "exit"; exit_code: number; error?: string };

export interface CreatePtyOptions {
  /** Full ws:// or wss:// URL including ?sandbox=&cols=&rows=. */
  url: string;
  /** Per-sandbox bearer token; sent in the Authorization header. */
  token?: string;
  /** Receives raw terminal output bytes as they arrive. */
  onData: (data: Uint8Array) => void;
}

/** A live interactive terminal handle. */
export class Pty {
  private readonly ws: WebSocket;
  private readonly onData: (data: Uint8Array) => void;
  private exitCode: number | undefined;
  private readonly exited: Promise<number>;
  private resolveExit!: (code: number) => void;

  constructor(ws: WebSocket, onData: (data: Uint8Array) => void) {
    this.ws = ws;
    this.onData = onData;
    this.exited = new Promise<number>((resolve) => {
      this.resolveExit = resolve;
    });
    this.ws.on("message", (raw: WebSocket.RawData) => this.handleMessage(raw));
    this.ws.on("close", () => {
      if (this.exitCode === undefined) {
        this.exitCode = -1;
        this.resolveExit(-1);
      }
    });
  }

  private handleMessage(raw: WebSocket.RawData): void {
    const text = raw.toString();
    const frame = JSON.parse(text) as PtyFrame;
    if (frame.kind === "output") {
      this.onData(b64ToBytes(frame.data));
    } else if (frame.kind === "exit") {
      this.exitCode = frame.exit_code;
      this.resolveExit(frame.exit_code);
      this.ws.close();
    }
  }

  /** Send raw keystroke bytes to the shell. */
  sendInput(data: Uint8Array): void {
    const frame: PtyFrame = { kind: "input", data: bytesToB64(data) };
    this.ws.send(JSON.stringify(frame));
  }

  /** Resize the terminal (TIOCSWINSZ in the guest, then SIGWINCH). */
  resize(cols: number, rows: number): void {
    const frame: PtyFrame = { kind: "resize", cols, rows };
    this.ws.send(JSON.stringify(frame));
  }

  /** Force-close; the guest kills the shell process group on disconnect. */
  kill(): void {
    this.ws.close();
    if (this.exitCode === undefined) {
      this.exitCode = -1;
      this.resolveExit(-1);
    }
  }

  /** Resolve with the shell exit code (or -1 if the connection dropped before
   * a terminal exit frame). */
  wait(): Promise<number> {
    return this.exited;
  }
}

/** Open a PTY WebSocket and resolve once it is open. The bearer token rides the
 * Authorization request header (supported by the Node `ws` client), so forkd's
 * ptyAuth gate sees it on the upgrade. */
export function createPty(opts: CreatePtyOptions): Promise<Pty> {
  return new Promise((resolve, reject) => {
    const headers: Record<string, string> = {};
    if (opts.token) {
      headers["Authorization"] = `Bearer ${opts.token}`;
    }
    const ws = new WebSocket(opts.url, ["mitos.pty.v1"], { headers });
    const pty = new Pty(ws, opts.onData);
    ws.on("open", () => resolve(pty));
    ws.on("error", (err: Error) =>
      reject(new Error(`pty websocket error: ${err.message}`)),
    );
  });
}

function bytesToB64(bytes: Uint8Array): string {
  let bin = "";
  for (const b of bytes) {
    bin += String.fromCharCode(b);
  }
  return btoa(bin);
}

function b64ToBytes(b64: string): Uint8Array {
  if (!b64) {
    return new Uint8Array(0);
  }
  const bin = atob(b64);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) {
    out[i] = bin.charCodeAt(i);
  }
  return out;
}
